packages/eslint-plugin/tests/rules/prefer-optional-chain/prefer-optional-chain.test.ts

```diff
@@ -1,1274 +1,2134 @@
+      },
+      {
+        code: noFormat`((foo ** 2) || {}).bar;`,
+        errors: [
+          {
+            messageId: 'preferOptionalChain',
+            column: 1,
+            endColumn: 23,
+            suggestions: [
+              {
+                messageId: 'optionalChainSuggest',
+                output: '(foo ** 2)?.bar;',
+              },
+            ],
+          },
+        ],
+      },
+      {
+        code: '(foo ** 2 || {}).bar;',
+        errors: [
+          {
+            messageId: 'preferOptionalChain',
+            column: 1,
+            endColumn: 21,
+            suggestions: [
+              {
+                messageId: 'optionalChainSuggest',
+                output: '(foo ** 2)?.bar;',
+              },
+            ],
+          },
+        ],
+      },
+      {
+        code: '(foo++ || {}).bar;',
+        errors: [
+          {
+            messageId: 'preferOptionalChain',
+            column: 1,
+            endColumn: 18,
+            suggestions: [
+              {
+                messageId: 'optionalChainSuggest',
+                output: '(foo++)?.bar;',
+              },
+            ],
+          },
+        ],
+      },
+      {
+        code: '(+foo || {}).bar;',
+        errors: [
+          {
+            messageId: 'preferOptionalChain',
+            column: 1,
+            endColumn: 17,
+            suggestions: [
+              {
+                messageId: 'optionalChainSuggest',
+                output: '(+foo)?.bar;',
+              },
+            ],
+          },
+        ],
+      },
+      {
+        code: '(this || {}).foo;',
+        errors: [
+          {
+            messageId: 'preferOptionalChain',
+            suggestions: [
+              {
+                messageId: 'optionalChainSuggest',
+                output: 'this?.foo;',
+              },
+            ],
+          },
+        ],
+      },
+    ],
+  });
+});
+
+describe('hand-crafted cases', () => {
+  ruleTester.run('prefer-optional-chain', rule, {
+    valid: [
+      '!a || !b;',
+      '!a || a.b;',
+      '!a && a.b;',
+      '!a && !a.b;',
+      '!a.b || a.b?.();',
+      '!a.b || a.b();',
+      'foo ||= bar;',
+      'foo ||= bar?.baz;',
+      'foo ||= bar?.baz?.buzz;',
+      'foo && bar;',
+      'foo && foo;',
+      'foo || bar;',
+      'foo ?? bar;',
+      'foo || foo.bar;',
+      'foo ?? foo.bar;',
+      "file !== 'index.ts' && file.endsWith('.ts');",
+      'nextToken && sourceCode.isSpaceBetweenTokens(prevToken, nextToken);',
+      'result && this.options.shouldPreserveNodeMaps;',
+      'foo && fooBar.baz;',
+      'match && match$1 !== undefined;',
+      "typeof foo === 'number' && foo.toFixed();",
+      "foo === 'undefined' && foo.length;",
+      'foo == bar && foo.bar == null;',
+      'foo === 1 && foo.toFixed();',
+      // call arguments are considered
+      'foo.bar(a) && foo.bar(a, b).baz;',
+      // type parameters are considered
+      'foo.bar<a>() && foo.bar<a, b>().baz;',
```

Comment:
`type parameters` were renamed to type arguments in this context, right?

Reply:
<img src="https://media1.giphy.com/media/1hMk0bfsSrG32Nhd5K/giphy.gif"/>